### PR TITLE
Rework UART module to follow `Extended API Guideline` and update documents

### PR DESCRIPTION
**Pull request by hs0225**
Rework UART module to follow `Extended API Guideline` and update documents 
 - Tested on raspberryPi3, stm32f4-discovery
 - Updated UART document

IoT.js-DCO-1.0-Signed-off-by: Hosung Kim hs852.kim@samsung.com

### CODE REVIEW

File: docs/api/IoT.js-API-UART.md [hunk 1]
```diff
@@ -1,46 +1,90 @@
...
+
+**Example**
 ```javascript
-var uart = require('uart');
-var options = {
+var Uart = require('uart');
+
+var uart = new Uart();
+var configuration = {
+  device: '/dev/ttyUSB0'
```
Comment (tadziopazur) [inline docs/api/IoT.js-API-UART.md]: Why does the application need to know these kind of details -- what devices might be there?
I'd rather the app could:
 - query what serial devices are there
 - be notified when serial devices come up/disappear
I do not necessarily think apps need to know serial port names (/dev/ttyUSB0 in this case).

Would this work on nuttx as well?

File: docs/api/IoT.js-API-UART.md [hunk 2]
```diff
@@ -1,46 +1,90 @@
...
+
+
+**Example**
 ```javascript
-var uart = require('uart');
-var options = {
+var Uart = require('uart');
+
+var uart = new Uart();
+var configuration = {
```
Comment (tadziopazur) [inline docs/api/IoT.js-API-UART.md]: Are these all the fields that can be set?
I'd like to know what options are there, and what options are supported by this particular port (i.e. speed range, CTS/RTS, XON/XOFF, ...).

File: docs/api/IoT.js-API-GPIO.md [hunk 3]
```diff
@@ -37,7 +37,7 @@ The following shows GPIO module APIs available for each platform.
 
 * On Linux, pin number is logical number starts from 1. Thus logical pin number *k* is not necessarily bound to physical pin number *k* in your board.
 * On Nuttx, you have to know pin number. The pin number is defined in target board module. For more module information, please see below list.
- * [STM32F4-discovery](../targets/nuttx-stm32f4/Stm32f4dis-Module.md).
+ * [STM32F4-discovery](../targets/nuttx-stm32f4/Stm32f4dis.md#gpio-pin).
```
Comment (glistening) [inline docs/api/IoT.js-API-GPIO.md]: 1) Please make the list indented. (I think it need one more spaces)
2) The link is broken. Please check it.
3) Remove . in the end of the line. It is not a sentence.

File: docs/api/IoT.js-API-PWM.md [hunk 4]
```diff
@@ -17,7 +17,7 @@ The following shows PWM module APIs available for each platform.
 ### `pin`
 * On Linux, `pin` is a pwm number which is 0 or 1.
 * On Nuttx, you have to know pin name. The pin name is defined in target board module. For more module information, please see below list.
- * [STM32F4-discovery](../../targets/nuttx-stm32f4/Stm32f4dis-Module.md).
+ * [STM32F4-discovery](../../targets/nuttx-stm32f4/Stm32f4dis.md#pwm-pin).
```
Comment (glistening) [inline docs/api/IoT.js-API-PWM.md]: Ditto. Please apply above comment 1, 3. ( The link is correct. )

File: docs/api/IoT.js-API-UART.md [hunk 5]
```diff
@@ -1,46 +1,90 @@
...
+## Contents
+  * [UART](#uart)
+    * [Constructor](#constructor)
+      * [`new UART()`](#new-uart)
+    * [Prototype Methods](#uart-prototype-methods)
+      * [`uart.open(configurable[, callback])`](#uart-open)
+  * [UARTPort](#uartport)
+    * [Prototype Methods](#uartport-prototype-methods)
+      * [`uartport.write(data[, callback])`](#uartport-read)
+      * [`uartport.writeSync(data)`](#uartport-read-sync)
```
Comment (glistening) [inline docs/api/IoT.js-API-UART.md]: Please make link to uartport-write and uartport-write-sync.

File: docs/api/IoT.js-API-UART.md (hunk 2 above)
Comment (hs0225) [inline docs/api/IoT.js-API-UART.md]: You are right. However, I have not implemented UART APIs and options (I rework UART module to follow 'Extended API Guideline'. I have updated the PR comment). APIs can be added to other PR later.

File: docs/api/IoT.js-API-UART.md (hunk 1 above)
Comment (hs0225) [inline docs/api/IoT.js-API-UART.md]: Most of our APIs come from the node module. Many node modules of system IO specify the device path. Your suggestion is good, but I've decided that this method will be more familiar to node users.

**Comment by hs0225** [comment]
@tadziopazur I updated PR as your comment. Thank you for your review.

File: docs/api/IoT.js-API-GPIO.md (hunk 3 above)
Comment (hs0225) [inline docs/api/IoT.js-API-GPIO.md]: I fixed it.

File: docs/api/IoT.js-API-PWM.md (hunk 4 above)
Comment (hs0225) [inline docs/api/IoT.js-API-PWM.md]: done.

File: docs/api/IoT.js-API-UART.md (hunk 5 above)
Comment (hs0225) [inline docs/api/IoT.js-API-UART.md]: done.

File: targets/nuttx-stm32f4/Stm32f4dis.md [hunk 6]
```diff
@@ -94,3 +94,34 @@ The following is a list of ADC pin name.
...
+| :--- | :---: |
+| USART2_RX | PA3 |
+| USART2_TX | PA2 |
+| USART3_RX | PB11 |
+| USART3_TX | PB10 |
+| UART5_RX | PD2 |
+| UART5_TX | PC12 |
+| USART6_RX | PC7 |
+| USART6_TX | PC6 |
+
```
Comment (glistening) [inline targets/nuttx-stm32f4/Stm32f4dis.md]: The table is broken.

File: test/run_pass/test_uart.js [hunk 7]
```diff
@@ -35,18 +36,21 @@ var serial = new uart(device, options, function(err){
...
-      serial.close();
+      serial.closeSync();
+      console.log('close done');
     }
   });
 
+  serial.writeSync("Hello IoT.js.\n\r");
 
-  serial.write("Hello there?\n", function(err) {
+  serial.write("Hello there?\n\r", function(err) {
```
Comment (glistening) [inline test/run_pass/test_uart.js]: Could you explain why you need to append '\r'? Is it for osx?

File: targets/nuttx-stm32f4/Stm32f4dis.md (hunk 6 above)
Comment (hs0225) [inline targets/nuttx-stm32f4/Stm32f4dis.md]: I fixed it.

File: test/run_pass/test_uart.js (hunk 7 above)
Comment (hs0225) [inline test/run_pass/test_uart.js]: It has nothing to do with osx. Characters can not be printed in the first column without `\r` in serial.

**Comment by yichoi** [comment]
LGTM

**Pull request merged**